[EDNA-76] DB concurrency fixes

worm2fed opened the pull request:
## Description

<!--
Describes the nature of your changes. If they are substantial, you should
further subdivide this into a section describing the problem you are solving and
another describing your solution.
-->

Problem:
1. When uploadFile is called, we add various data to DB using multiple INSERT statements. We put them into transact. My idea was to achieve atomicity by this transact, but apparently it doesn't work the way I want. For example: add threadDelay $ 30 * 1000 * 1000 after insertTarget and try to upload a file. Once you click "Add experiments", go to the library. You'll see targets there, but no compounds (yet). AFAIU it means that other threads can observe the state when a file is only partially uploaded. For example, it's possible that a sub-experiment is already added, but removed measurements are not.
2. Extension of the previous one: try killing edna-server with SIGINT during threadDelay $ 30 * 1000 * 1000. If you run it again, you'll see that targets have been added and can be seen, but the rest is not. It's a more serious problem than (1) because (1) normally may last only for ≈a second, but this problem is persistent.

The root of problem:
We have three functions: `withConnection`, `transact`  and `runPg`. Function `transact` supposed to wrap some code inside transactions, and it’s working almost as expected. If we look at postgresql log we can see that transactions starts with process id `X` and ends with same one, but actions inside transaction has another id `Y`.
So, the problem is that when we call transact we use withConnection and particular actions uses withConnection too, which produces different processes.

Solution:
Store functions to work with connections in `EdnaContext` instead of `ConnPoll`, and then when we need to execute some actions in transaction - override that functions to use single connection.

## Related issue(s)

<!--
- Short description of how the PR relates to the issue, including an issue link.
For example
- Fixed #100500 by adding lenses to exported items

Write 'None' if there are no related issues (which is discouraged).
-->

Resolves https://issues.serokell.io/issue/EDNA-76

## :white_check_mark: Checklist for your Pull Request

<!--
Ideally a PR has all of the checkmarks set.

If something in this list is irrelevant to your PR, you should still set this
checkmark indicating that you are sure it is dealt with (be that by irrelevance).

If you don't set a checkmark (e. g. don't add a test for new functionality),
you must be able to justify that.
-->

#### Related changes (conditional)

- Tests
  - [ ] If I added new functionality, I added tests covering it.
  - [x] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

- Documentation
  - [x] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)

#### Stylistic guide (mandatory)

- [x] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Commit-and-PR-policy-4cf98e1a910a415d86b5f2491d9af1af).
- [x] My code complies with the [style guide](../tree/master/docs/code-style.md).

worm2fed committed:
Ignore python cache

Problem:
We have no gitignore for analysis directory, and somebody can add python cache to repo, we do not want it.

Solution:
Simply add `__pycache__/` to ignore.

worm2fed committed:
Fix transactions execution

Problem:
We have three functions: `withConnection`, `transact`  and `runPg`. Function `transact` supposed to wrap some code inside transactions, and it’s working almost as expected. If we look at postgresql log we can see that transactions starts with process id `X` and ends with same one, but actions inside transaction has another id `Y`.
So, the problem is that when we call transact we use withConnection and particular actions uses withConnection too, which produces different processes.

Solution:
Store functions to work with connections in `EdnaContext` instead of `ConnPoll`, and then when we need to execute some actions in transaction - override that functions to use single connection.

worm2fed commented:
@gromakovsky I've fixed the main problem.

In task description we additionally have following:
> There are some functions (e. g. updateProject) where we read some data from DB and then update DB depending on this data. The problem is that some action may happen between read and update and cause uncaught SQL error (which will be shown as Internal Server Error which is not ok) or maybe even invalid state (if we don't check something by SQL constraint). For example, in case of updateProject 2 people may concurrently change name of 2 different projects to X. One of these calls may fail with SqlError in this case. One possible solution is to do UPDATE in one query with subquery that checks that no projects with such name exist.

So, I think it's ok for now. We have mechanism to catch SqlError when there name duplications, and then we throw err400. 
Of course we can wrap everything in transactions, then if one user starts updating project and another one wants to update project too - he shall wait until first one commits transactions or rolls it back.
It described more detailed here: https://wiki.postgresql.org/wiki/SSI

gromakovsky commented on backend/src/Edna/DB/Connection.hs:
```diff
@@ -63,3 +67,42 @@ createConnPool (ConnString connStr) maxConnsNum = liftIO $ ConnPool <$>
+
+-- | Provides methods for working with connections to Postgres database.
+data PostgresConn = PostgresConn
+  { _withConnection  :: forall (m :: Type -> Type) a. MonadUnliftIO m
```
According to our general convention I think it's better to call it `_pcWithConnection`. Also is `_` necessary? I think it's usually used for somewhat private things or when there is another thing with the same name without `_`. But here these things don't look private in any sense and apparently don't conflict with anything (especially if we add `pc` prefix).

UPD: I see that `_withConnection` conflicts with `withConnection` and in this sense it's somewhat private indeed. So maybe `_` is justified, but I still think we can just replace it with `pc` prefix.

worm2fed replied:
I'll fix it

gromakovsky commented on backend/src/Edna/DB/Connection.hs:
```diff
@@ -63,3 +67,42 @@ createConnPool (ConnString connStr) maxConnsNum = liftIO $ ConnPool <$>
+                     => Connection -> m a  -> m a
+  }
+
+-- | Implementation of @PostgresConn@ which uses a @ConnPool@ to support
```
Here I think you can use single quotes `'` instead of `@` because `PostgresConn` and `ConnPool` are actual identifiers in code. And below you should write `@withConnection@` instead of ``` `withConnection` ``` because these are not real identifiers (or write `'_withConnection'` for example).

https://github.com/ghc/haddock/blob/master/doc/cheatsheet/haddocks.pdf

gromakovsky commented on backend/src/Edna/DB/Integration.hs:
```diff
@@ -30,21 +29,24 @@ import Database.Beam.Query
+  c <- view edDBConnection
+  _withConnection c $ \conn ->
+    _withTransaction c conn $ do
+      local (\ctx -> ctx { _edDBConnection = postgresConnSingle conn }) $ do
```
Might be simpler to write `let c' = postgresConnSingle conn`, then `ctx { _edDBConnection = c' }` here and just use `c'` below (without `view`). But that's up to you.

gromakovsky reviewed (changes requested):
The fix looks good, I have only cosmetic comments. I'm going to test it now.

gromakovsky commented:
> I'm going to test it now.

Alright, it seems to work correctly :clap:

gromakovsky commented:
I think it'd be nice to also add a test. We can cause an error inside `uploadFile'` somehow and check that nothing is added to DB in this case. For example, we can pass `error` as `fileName` to it, expect exception in the test and then check that no compounds and targets are added. If you do it this way, please also add a comment to `uploadFile'` saying that there is a test which relies on the fact that we add experiment file after compounds and targets.

gromakovsky commented:
> We have mechanism to catch SqlError when there name duplications, and then we throw err400.

AFAIU currently we don't explicitly catch `SqlError` anywhere, so it will be an internal server error in this case. We can catch `SqlError`, but I slightly dislike it because:
1. If we catch any `SqlError` and just return 400 in this case, we may catch an unexpected error that happens for a reason other than bad request. It's hard to predict what errors may happen when you make SQL query. For example, we can get an error if we change the DB schema and forget to migrate/wipe existing DB, in this case we shouldn't respond with 400. So ideally we should distinguish `SqlError`s.
2. However, if we want to distinguish `SqlError`s, we need to parse textual value attached to it and it doesn't sound very robust. I'm afraid that these textual values may change and they may even be different for different versions of PostgreSQL. It's unlikely, but still sounds possible.

Another problem is that some constraint may be not enforced at the level of SQL: maybe we forgot to add such a constraint or it's just too hard to express in terms of SQL. Then if we check it only in Haskell and then insert some data after the check, we may insert incorrect data if there was a change between the check and the insertion.

So I think ideally we should update functions like `updateProject` (which do `SELECT` and `UPDATE` or `INSERT`) to contain one transaction: either using `transact` or by grouping `SELECT` and `UPDATE` into one query (where it's possible).

worm2fed commented:
@gromakovsky I've got it.

> So I think ideally we should update functions like updateProject (which do SELECT and UPDATE or INSERT) to contain one transaction: either using transact or by grouping SELECT and UPDATE into one query (where it's possible).

What about `getProject`? In this case I think we should wrap it in `transact` too, because we one user can start updating and another one can request this project

worm2fed committed:
Wrap some library actions in transact

Problem:
There are some functions (e. g. updateProject) where we read some data from DB and then update DB depending on this data.
The problem is that some action may happen between read and update and cause uncaught SQL error (which will be shown as Internal Server Error which is not ok) or maybe even invalid state (if we don't check something by SQL constraint).
For example, in case of updateProject 2 people may concurrently change name of 2 different projects to X. One of these calls may fail with SqlError in this case.

Solution:
Wrap some library actions in transact to avoid undesired sql exceptions raised by concurrency requests.
For more information see https://wiki.postgresql.org/wiki/SSI.

gromakovsky commented:
> What about `getProject`? In this case I think we should wrap it in `transact` too, because we one user can start updating and another one can request this project

`getProject` looks ok, it makes just one query, there is no point in adding `transact` to it.

Oh, maybe you mean `addProject`, then yes, it makes sense there.

gromakovsky commented:
> I think it'd be nice to also add a test. 

Please do it and I'll approve.

worm2fed committed:
Add tests for upload file

Problem:
We want to be sure that if some error occurs due uploading file, created targets and compounds will be rolled back.

Solution:
Add test, which passes error instead of filename, then expects that error and then checks amount of targets, compounds and experiments in DB.

worm2fed commented:
@gromakovsky done

gromakovsky commented on backend/test/Test/UploadSpec.hs:
```diff
@@ -101,6 +107,19 @@ spec = withContext $ startWithInitial $ do
+      compounds <- runRIO ctx $ Library.getCompounds Nothing Nothing Nothing
+      experiments <- runRIO ctx $ Dashboard.getExperiments Nothing Nothing Nothing
+
+      -- We do not use 0 here, because we add some data in upper tests
```
Actually it's trickier, you shouldn't rely on which tests will run before. For example if you pass `--match rollback` to the test-suite, this test will fail because it will run only this test. When I was writing such tests I tried to carefully structure them, so that no matter what pattern we pass to `--match`, all tests will succeed. Specifically, it's better to put tests that don't mutate DB before mutating tests. In this case it should solve the problem, just move this test before `"successfully adds targets and compounds"` and test that all numbers below are 0.

If you do it, you can also define ```shouldBeZero l = liftIO $ length l `shouldBe` 0``` and use `runRIO ctx` once instead of 3 times:
```
runRIO ctx $ do
  Library.getTargets Nothing Nothing Nothing >>= shouldBeZero
  Library.getCompounds Nothing Nothing Nothing >>= shouldBeZero
  Dashboard.getExperiments Nothing Nothing Nothing >>= shouldBeZero
```

worm2fed replied:
That was first I've did. But unfortunately, after transaction rollback postgres increases auto_increment index.
I can put it in separate describe section with `startWithInitial`

worm2fed replied:
executing `_ <- ctx <$ runWithInit ctx addInitialData` before other actions inside tests works as well

gromakovsky replied:
Oh, indeed, there is a problem. Okay, `runWithInit ctx addInitialData` looks good to me.

gromakovsky commented on backend/test/Test/UploadSpec.hs:
```diff
@@ -101,6 +107,19 @@ spec = withContext $ startWithInitial $ do
+      -- We do not use 0 here, because we add some data in upper tests
+      length targets `shouldBe` 4
+      length compounds `shouldBe` 5
+      length (wItem <$> erExperiments experiments) `shouldBe` 10
```
Btw, `wItem <$>` here looks weird. I mean, length of `erExperiments experiments` must be equal to the length of `(wItem <$> erExperiments experiments)`, `<$>` applied to a list can't change its length.

gromakovsky commented on backend/test/Test/UploadSpec.hs:
```diff
@@ -101,6 +107,19 @@ spec = withContext $ startWithInitial $ do
+
+      -- We do not use 0 here, because we add some data in upper tests
+      length targets `shouldBe` 4
+      length compounds `shouldBe` 5
```
Also currently these checks don't test the actual problem, because even if you add `sampleFile` successfully, you won't add new compounds/targets because they are already added.

worm2fed replied:
agree

gromakovsky commented:
I reverted your "Fix transactions execution" commit and all tests still pass, can you please check why?

worm2fed commented:
I'll check it

gromakovsky commented on backend/test/Test/UploadSpec.hs:
```diff
@@ -109,17 +109,15 @@ spec = withContext $ startWithInitial $ do
         onlyDisabled secondaryMeasurements `shouldBe` secondaryDisabled
 
     it "rollbacks targets and compounds due processing error" $ \ctx -> do
+      _ <- ctx <$ runWithInit ctx addInitialData
```
`ctx <$ ` causes this expression to return `ctx` and you ignore it. So `ctx <$` is redundant.

worm2fed commented:
> I reverted your "Fix transactions execution" commit and all tests still pass, can you please check why?

the problem was that tests was running with 'dirty' database, after fixes with tests it fail (in case of revert).
<img width="767" alt="image" src="https://user-images.githubusercontent.com/6579275/116701030-1d9c3400-a9d0-11eb-80f1-1a9f654848eb.png">

gromakovsky reviewed (approved):
LGTM

gromakovsky committed:
Improve error message in our of the test

Problem: if something goes wrong in Library test, `Map.!` operator may
fail and the error message in this case is not very helpful:
```
       uncaught exception: ErrorCall
       Map.!: given key is not an element in the map
       CallStack (from HasCallStack):
         error, called at libraries/containers/containers/src/Data/Map/Internal.hs:627:17 in containers-0.6.2.1:Data.Map.Internal
```
It refers to some library file and doesn't say what key is not found.

Solution: if key is not found, fail using `expectationFailure` and
attach the key to the error.

worm2fed committed:
Fix target getter tests

Problem:
After transaction fix one test is failing. The reason is that
in transaction all targets for now have same creation date,
so then in sorting by date there are no difference between
first three targets. We expect 5, 3, 2 and request returns 5, 2, 1.

Solution:
Expect only one target (drop others). As we can see, only first three
targets has exactly same time (1, 2, 3), but we have another one with
id 5 which was added with another transaction and hence has another
date. The most simple solution we can apply here (note, that we use
desc sorting) is just limit result to one target and expect that we'll
receive target with id 5.
